hitch/hitch_1.7.3.patch

```diff
@@ -0,0 +1,237 @@
++		[wolfSSL has SSL_CTX_get_default_passwd_cb_userdata()])
++	AC_DEFINE([OPENSSL_WITH_LOCKS], [1], [wolfSSL needs explicit locking])
++	AC_DEFINE([HAVE_X509_NAME_ENTRY_GET_DATA], [1],
++		[wolfSSL has X509_NAME_ENTRY_get_data()])
++	AC_DEFINE([HAVE_X509_STORE_GET0_OBJECTS], [1],
++		[wolfSSL has X509_STORE_get0_objects()])
++	AC_DEFINE([HAVE_X509_OBJECT_GET0_X509], [1],
++		[wolfSSL has X509_OBJECT_get0_X509()])
++
++		HITCH_CHECK_FLAGS([HITCH_CFLAGS], [
++		-I$wolfssl_path/include,
++		-I$wolfssl_path/include/wolfssl,
++		-DEXTERNAL_OPTS_OPENVPN
++	])
++else
+ 	HITCH_CHECK_FUNC([SSL_get0_alpn_selected], [$SSL_LIBS], [
+ 		AC_DEFINE([OPENSSL_WITH_ALPN], [1], [OpenSSL supports ALPN])
+ 	])
+@@ -223,6 +262,7 @@ AS_VERSION_COMPARE([$($PKG_CONFIG --modversion openssl)], [1.1.1],
+ 
+ 	AS_IF([test "x$openssl111" = xyes],
+ 		[AC_DEFINE([HAVE_TLS_1_3], [1], [Define to 1 if TLSv1.3 is available])], [])
++fi
+ 
+ SH_TESTS="$(cd $srcdir/src && echo tests/test*.sh)"
+ AC_SUBST(SH_TESTS)
+diff --git a/src/hitch.c b/src/hitch.c
+index 7d15207..071b017 100644
+--- a/src/hitch.c
++++ b/src/hitch.c
+@@ -36,7 +36,6 @@
+   */
+ 
+ #include "config.h"
+-
+ #include <openssl/x509.h>
+ #include <openssl/x509_vfy.h>
+ #include <openssl/engine.h>
+@@ -1325,6 +1324,7 @@ init_openssl(void)
+ 	SSL_load_error_strings();
+ 	OpenSSL_add_all_digests();
+ 
++#ifndef WITH_WOLFSSL
+ 	if (CONFIG->ENGINE) {
+ 		ENGINE *e = NULL;
+ 		ENGINE_load_builtin_engines();
+@@ -1345,6 +1345,7 @@ init_openssl(void)
+ 			ENGINE_free(e);
+ 		}
+ 	}
++#endif
+ }
+ 
+ static void
+@@ -2034,7 +2035,7 @@ proxy_tlv_cert(struct proxystate *ps, char *dst, ssize_t dstlen)
+ {
+ 	X509 *crt;
+ 	BIO *bio;
+-	struct buf_mem_st bm[1];
++	BUF_MEM bm[1];
+ 
+ 	crt = SSL_get_peer_certificate(ps->ssl);
+ 	if (crt == NULL)
+diff --git a/src/ocsp.c b/src/ocsp.c
+index 8f5639a..437734b 100644
+--- a/src/ocsp.c
++++ b/src/ocsp.c
+@@ -547,7 +547,12 @@ hocsp_query_responder(struct ev_loop *loop, ev_timer *w, int revents)
+ {
+ 	ocspquery *oq;
+ 	OCSP_REQUEST *req = NULL;
++#ifndef WITH_WOLFSSL
+ 	OCSP_REQ_CTX *rctx = NULL;
++#else
++	unsigned char *httpBuf = NULL, *reqBuf = NULL, *respBuf = NULL;
++	int httpBufSz = 4096, reqSz = 0, respSz = 4096;
++#endif
+ 	STACK_OF(OPENSSL_STRING) *sk_uri;
+ 	char *host = NULL, *port = NULL, *path = NULL;
+ 	int https = 0;
+@@ -647,6 +652,30 @@ hocsp_query_responder(struct ev_loop *loop, ev_timer *w, int revents)
+ 		}
+ 	}
+ 
++#ifdef WITH_WOLFSSL
++	/* Get request size */
++	reqSz = wolfSSL_i2d_OCSP_REQUEST(req, NULL);
++	if (reqSz < 0) {
++		goto err;
++	}
++
++	reqBuf = (unsigned char*) malloc(reqSz);
++	if (reqBuf == NULL) {
++		goto err;
++	}
++	n = wolfSSL_i2d_OCSP_REQUEST(req, &reqBuf);
++	if (n < 0) {
++		goto err;
++	}
++	httpBuf = (unsigned char*) malloc(httpBufSz);
++	if (httpBuf == NULL) {
++		goto err;
++	}
++	httpBufSz = wolfIO_HttpBuildRequestOcsp(host, path, n, httpBuf, httpBufSz);
++	if (httpBufSz <= 0) {
++		goto err;
++	}
++#else
```

Comment:
Do you think it would a lot of effort to implement the `OCSP_*` functions in wolfSSL?

Reply:
I believe it would be a decent effort, as they differ significantly from how wolfSSL handles OCSP.  I also felt that these are niche APIs, so I didn't feel they were worth the effort to add when it was straightforward to just rewrite hitch's code.
Let me know if you disagree, and I can reconsider implementing them.

Reply:
If these API show up elsewhere we will reconsider. For now the workaround is fine.